File: src/core/schema_tests/cpp/test_orbbec_camera.cpp
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

#include <catch2/catch_test_macros.hpp>
#include <flatbuffers/flatbuffers.h>
#include <schema/orbbec_audio_generated.h>
#include <schema/orbbec_calibration_generated.h>
#include <schema/orbbec_camera_generated.h>
#include <schema/orbbec_device_state_generated.h>
#include <schema/orbbec_imu_generated.h>
#include <schema/timestamp_generated.h>

#include <cstdint>
#include <memory>

#define VT(field) (field + 2) * 2

static_assert(core::FrameMetadataOrbbec::VT_STREAM == VT(0));
static_assert(core::FrameMetadataOrbbec::VT_SEQUENCE_NUMBER == VT(1));
static_assert(core::FrameMetadataOrbbec::VT_WIDTH == VT(2));
static_assert(core::FrameMetadataOrbbec::VT_HEIGHT == VT(3));
static_assert(core::FrameMetadataOrbbec::VT_FPS == VT(4));
static_assert(core::FrameMetadataOrbbec::VT_PIXEL_FORMAT == VT(5));
static_assert(core::FrameMetadataOrbbec::VT_ENCODED_BYTES == VT(6));
static_assert(core::FrameMetadataOrbbec::VT_SDK_METADATA == VT(7));
static_assert(core::FrameMetadataOrbbec::VT_CAPTURE_EPOCH == VT(8));
static_assert(core::OrbbecPixelFormat_Mjpg == 0);

TEST_CASE("Orbbec camera metadata round trips", "[orbbec][schema]")
{
    core::FrameMetadataOrbbecT original;
    original.stream = core::OrbbecCameraStream_ColorRight;
    original.sequence_number = UINT64_MAX;
    original.width = 1280;
    original.height = 720;
    original.fps = 30;
    original.pixel_format = core::OrbbecPixelFormat_Mjpg;
    original.encoded_bytes = 1234;
    original.sdk_metadata.emplace_back(1, 99);
    original.capture_epoch = 3;

    flatbuffers::FlatBufferBuilder builder;
    builder.Finish(core::FrameMetadataOrbbec::Pack(builder, &original));

    core::FrameMetadataOrbbecT restored;
    flatbuffers::GetRoot<core::FrameMetadataOrbbec>(builder.GetBufferPointer())->UnPackTo(&restored);
    CHECK(restored.stream == core::OrbbecCameraStream_ColorRight);
    CHECK(restored.sequence_number == UINT64_MAX);
    CHECK(restored.width == 1280);
    CHECK(restored.height == 720);
    CHECK(restored.fps == 30);
    CHECK(restored.pixel_format == core::OrbbecPixelFormat_Mjpg);
    CHECK(restored.encoded_bytes == 1234);
    REQUIRE(restored.sdk_metadata.size() == 1);
    CHECK(restored.sdk_metadata[0].value() == 99);
    CHECK(restored.capture_epoch == 3);
}

TEST_CASE("Orbbec Ego auxiliary schemas round trip", "[orbbec][schema]")
{
    core::OrbbecImuBatchT imu;
    imu.sensor = core::OrbbecImuSensor_Gyro;
    imu.sequence_number = 5;
    imu.sample_rate_hz = 1000;
    imu.full_scale = 500;
    imu.samples.emplace_back(1, 2, 3, 24, 100, 200);
    imu.capture_epoch = 2;
    flatbuffers::FlatBufferBuilder imu_builder;
    imu_builder.Finish(core::OrbbecImuBatch::Pack(imu_builder, &imu));
    core::OrbbecImuBatchT restored_imu;
    flatbuffers::GetRoot<core::OrbbecImuBatch>(imu_builder.GetBufferPointer())->UnPackTo(&restored_imu);
    REQUIRE(restored_imu.samples.size() == 1);
    CHECK(restored_imu.samples[0].z_si() == 3);
    CHECK(restored_imu.capture_epoch == 2);

    core::OrbbecAudioChunkT audio;
    audio.sample_rate_hz = 48000;
    audio.channel_count = 1;
    audio.bits_per_sample = 16;
    audio.sample_count = 480;
    audio.wav_data_offset = 44;
    audio.capture_epoch = 2;
    flatbuffers::FlatBufferBuilder audio_builder;
    audio_builder.Finish(core::OrbbecAudioChunk::Pack(audio_builder, &audio));
    core::OrbbecAudioChunkT restored_audio;
    flatbuffers::GetRoot<core::OrbbecAudioChunk>(audio_builder.GetBufferPointer())->UnPackTo(&restored_audio);
    CHECK(restored_audio.wav_data_offset == 44);
    CHECK(restored_audio.sample_count == 480);
    CHECK(restored_audio.capture_epoch == 2);

    core::OrbbecCalibrationT calibration;
    calibration.device_uid = "ego";
    calibration.raw_alignment_yaml = "camera: left";
    calibration.capture_epoch = 2;
    flatbuffers::FlatBufferBuilder calibration_builder;
    calibration_builder.Finish(core::OrbbecCalibration::Pack(calibration_builder, &calibration));
    core::OrbbecCalibrationT restored_calibration;
    flatbuffers::GetRoot<core::OrbbecCalibration>(calibration_builder.GetBufferPointer())->UnPackTo(&restored_calibration);
    CHECK(restored_calibration.device_uid == "ego");
    CHECK(restored_calibration.capture_epoch == 2);

    core::OrbbecDeviceStateT state;
    state.sequence_number = 7;
    state.temperature_c = 31.5f;
    state.capture_health = core::OrbbecCaptureHealth_Warning;
    state.failure_reason = "queue warning";
    state.queue_capacity = 4096;
    state.queue_peak = 3482;
    state.capture_epoch = 2;
    state.connection_state = core::OrbbecConnectionState_Recovered;
    state.reconnect_attempt = 4;
    state.properties.emplace_back(279, 8'000'000);
    flatbuffers::FlatBufferBuilder state_builder;
    state_builder.Finish(core::OrbbecDeviceState::Pack(state_builder, &state));
    core::OrbbecDeviceStateT restored_state;
    flatbuffers::GetRoot<core::OrbbecDeviceState>(state_builder.GetBufferPointer())->UnPackTo(&restored_state);
    CHECK(restored_state.temperature_c == 31.5f);
    CHECK(restored_state.capture_health == core::OrbbecCaptureHealth_Warning);
    CHECK(restored_state.failure_reason == "queue warning");
    CHECK(restored_state.queue_capacity == 4096);
    CHECK(restored_state.capture_epoch == 2);
    CHECK(restored_state.connection_state == core::OrbbecConnectionState_Recovered);
    CHECK(restored_state.reconnect_attempt == 4);
    REQUIRE(restored_state.properties.size() == 1);
}

TEST_CASE("Orbbec camera record retains timestamp", "[orbbec][schema]")
{
    auto record = std::make_shared<core::FrameMetadataOrbbecRecordT>();
    record->data = std::make_shared<core::FrameMetadataOrbbecT>();
    record->timestamp = std::make_shared<core::DeviceDataTimestamp>(1, 2, 3);

    flatbuffers::FlatBufferBuilder builder;
    builder.Finish(core::FrameMetadataOrbbecRecord::Pack(builder, record.get()));
    const auto* restored = flatbuffers::GetRoot<core::FrameMetadataOrbbecRecord>(builder.GetBufferPointer());
    REQUIRE(restored->timestamp() != nullptr);
    CHECK(restored->timestamp()->available_time_local_common_clock() == 1);
    CHECK(restored->timestamp()->sample_time_local_common_clock() == 2);
    CHECK(restored->timestamp()->sample_time_raw_device_clock() == 3);
}

TEST_CASE("Orbbec embedded media schemas round trip", "[orbbec][schema]")
{
    core::OrbbecEncodedVideoFrameT video;
    video.stream = core::OrbbecCameraStream_ColorRight;
    video.sequence_number = 42;
    video.width = 1600;
    video.height = 1300;
    video.fps = 30;
    video.pixel_format = core::OrbbecPixelFormat_H264;
    video.encoded_data = { 0, 0, 0, 1, 0x65, 1 };
    video.capture_epoch = 2;
    flatbuffers::FlatBufferBuilder video_builder;
    video_builder.Finish(core::OrbbecEncodedVideoFrame::Pack(video_builder, &video));
    core::OrbbecEncodedVideoFrameT restored_video;
    flatbuffers::GetRoot<core::OrbbecEncodedVideoFrame>(video_builder.GetBufferPointer())->UnPackTo(&restored_video);
    CHECK(restored_video.stream == core::OrbbecCameraStream_ColorRight);
    CHECK(restored_video.encoded_data == video.encoded_data);
    CHECK(restored_video.capture_epoch == 2);

    core::OrbbecPcmAudioChunkT audio;
    audio.sequence_number = 7;
    audio.sample_rate_hz = 48000;
    audio.channel_count = 1;
    audio.bits_per_sample = 16;
    audio.sample_count = 2;
    audio.pcm_data = { 1, 0, 2, 0 };
    audio.capture_epoch = 2;
    flatbuffers::FlatBufferBuilder audio_builder;
    audio_builder.Finish(core::OrbbecPcmAudioChunk::Pack(audio_builder, &audio));
    core::OrbbecPcmAudioChunkT restored_audio;
    flatbuffers::GetRoot<core::OrbbecPcmAudioChunk>(audio_builder.GetBufferPointer())->UnPackTo(&restored_audio);
    CHECK(restored_audio.pcm_data == audio.pcm_data);
    CHECK(restored_audio.capture_epoch == 2);
}

File: src/core/schema_tests/python/test_orbbec_camera.py
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

from isaacteleop.schema import (
    DeviceDataTimestamp,
    FrameMetadataOrbbec,
    FrameMetadataOrbbecRecord,
    OrbbecCameraStream,
    OrbbecPixelFormat,
    OrbbecAudioChunk,
    OrbbecPcmAudioChunk,
    OrbbecCalibration,
    OrbbecCaptureHealth,
    OrbbecConnectionState,
    OrbbecDevicePropertyValue,
    OrbbecDeviceState,
    OrbbecFrameMetadataEntry,
    OrbbecImuBatch,
    OrbbecImuSample,
    OrbbecImuSensor,
    OrbbecEncodedVideoFrame,
)


def test_orbbec_metadata_defaults_and_properties():
    metadata = FrameMetadataOrbbec()
    assert metadata.stream == OrbbecCameraStream.ColorLeft
    assert metadata.pixel_format == OrbbecPixelFormat.Mjpg

    metadata.stream = OrbbecCameraStream.ColorRight
    metadata.sequence_number = 2**64 - 1
    metadata.width = 1280
    metadata.height = 720
    metadata.fps = 30
    metadata.pixel_format = OrbbecPixelFormat.H265
    metadata.encoded_bytes = 123
    metadata.capture_epoch = 3
    metadata.sdk_metadata = [OrbbecFrameMetadataEntry(1, 99)]

    assert metadata.stream == OrbbecCameraStream.ColorRight
    assert metadata.sequence_number == 2**64 - 1
    assert "ColorRight" in repr(metadata)
    assert metadata.pixel_format == OrbbecPixelFormat.H265
    assert metadata.sdk_metadata[0].value == 99
    assert metadata.capture_epoch == 3


def test_orbbec_record_keeps_timestamp():
    metadata = FrameMetadataOrbbec()
    timestamp = DeviceDataTimestamp(1, 2, 3)
    record = FrameMetadataOrbbecRecord(metadata, timestamp)

    assert record.data is not None
    assert record.timestamp.sample_time_raw_device_clock == 3


def test_orbbec_ego_auxiliary_types():
    imu = OrbbecImuBatch()
    imu.sensor = OrbbecImuSensor.Gyro
    imu.sample_rate_hz = 1000
    imu.samples = [OrbbecImuSample(1, 2, 3, 24, 100, 200)]
    imu.capture_epoch = 2
    assert imu.samples[0].z_si == 3

    audio = OrbbecAudioChunk()
    audio.sample_rate_hz = 48000
    audio.wav_data_offset = 44
    audio.capture_epoch = 2
    assert audio.wav_data_offset == 44
    assert audio.capture_epoch == 2

    calibration = OrbbecCalibration()
    calibration.device_uid = "ego"
    calibration.capture_epoch = 2
    assert calibration.device_uid == "ego"
    assert calibration.capture_epoch == 2

    state = OrbbecDeviceState()
    state.capture_health = OrbbecCaptureHealth.Warning
    state.failure_reason = "queue warning"
    state.queue_capacity = 4096
    state.capture_epoch = 2
    state.connection_state = OrbbecConnectionState.Recovered
    state.reconnect_attempt = 4
    state.properties = [OrbbecDevicePropertyValue(279, 8_000_000)]
    assert state.properties[0].property_id == 279
    assert state.capture_health == OrbbecCaptureHealth.Warning
    assert state.connection_state == OrbbecConnectionState.Recovered
    assert state.reconnect_attempt == 4


def test_orbbec_embedded_media_types():
    video = OrbbecEncodedVideoFrame()
    video.stream = OrbbecCameraStream.ColorRight
    video.encoded_data = [0, 0, 0, 1, 0x65]
    video.capture_epoch = 2
    assert video.stream == OrbbecCameraStream.ColorRight
    assert video.encoded_data[-1] == 0x65
    assert video.capture_epoch == 2

    audio = OrbbecPcmAudioChunk()
    audio.sample_rate_hz = 48000
    audio.pcm_data = [1, 0, 2, 0]
    audio.capture_epoch = 2
    assert audio.pcm_data == [1, 0, 2, 0]
    assert audio.capture_epoch == 2
